packages/graphql-codegen-cli/src/utils/watcher.ts

```diff
@@ -136,32 +133,82 @@ export const createWatcher = (
 
-    const shutdown = () => {
+    const shutdown = (
+      /** Optional callback to execute after shutdown has completed its async tasks */
+      afterShutdown?: () => void
+    ) => {
       isShutdown = true;
       debugLog(`[Watcher] Shutting down`);
       log(`Shutting down watch...`);
-      watcherSubscription.unsubscribe();
-      lifecycleHooks(config.hooks).beforeDone();
+
+      const pendingUnsubscribe = watcherSubscription.unsubscribe();
+      const pendingBeforeDoneHook = lifecycleHooks(config.hooks).beforeDone();
+
+      if (afterShutdown && typeof afterShutdown === 'function') {
+        Promise.allSettled([pendingUnsubscribe, pendingBeforeDoneHook]).then(afterShutdown);
+      }
     };
 
-    process.once('SIGINT', shutdown);
-    process.once('SIGTERM', shutdown);
+    abortSignal.addEventListener('abort', () => shutdown(abortSignal.reason));
+
+    process.once('SIGINT', () => shutdown());
+    process.once('SIGTERM', () => shutdown());
+  };
+
+  // Use an AbortController for shutdown signals
+  // NOTE: This will be polyfilled on Node 14 (or any environment without it defined)
+  const abortController = new AbortController();
+
+  /**
+   * Send shutdown signal and return a promise that only resolves after the
+   * runningWatcher has resolved, which only resolved after the shutdown signal has been handled
+   */
+  const stopWatching = async function () {
+    // stopWatching.afterShutdown is lazily set to resolve pendingShutdown promise
+    abortController.abort(stopWatching.afterShutdown);
+
+    // SUBTLE: runningWatcher waits for pendingShutdown before it resolves itself, so
+    // by awaiting it here, we are awaiting both the shutdown handler, and runningWatcher itself
+    await stopWatching.runningWatcher;
+  };
+  stopWatching.afterShutdown = () => {
+    debugLog('Shutdown watcher before it started');
   };
+  stopWatching.runningWatcher = Promise.resolve();
+
+  /** Promise will resolve after the shutdown() handler completes */
+  const pendingShutdown = new Promise<void>(afterShutdown => {
+    // afterShutdown will be passed to shutdown() handler via abortSignal.reason
+    stopWatching.afterShutdown = afterShutdown;
+  });
 
-  // the promise never resolves to keep process running
-  return new Promise<void>((resolve, reject) => {
-    executeCodegen(initalContext)
+  /**
+   * Promise that resolves after the watch server has shutdown, either because
+   * stopWatching() was called or there was an error inside it
+   */
+  stopWatching.runningWatcher = new Promise<void>((resolve, reject) => {
+    executeCodegen(initialContext)
       .then(onNext, () => Promise.resolve())
-      .then(runWatcher)
+      .then(() => runWatcher(abortController.signal))
       .catch(err => {
         watcherSubscription.unsubscribe();
         reject(err);
+      })
+      .then(() => pendingShutdown)
+      .finally(() => {
+        debugLog('Done watching.');
+        resolve();
       });
   });
```

Comment:
There are some subtleties to this shutdown order, because the two promises rely on each other and there is some lazy property assignment. The gist of it is that `await stopWatching()` will only resolve after the `runningWatcher` has resolved, and the intermediate step of `runningWatcher` will wait for `pendingShutdown`, which is ultimately the callback passed to the `shutdown` handler, to complete.

I tested this extensively with a bunch of log messages, and every time I ran tests, I passed `--detectOpenHandles`, so I'm very confident that this shutdown logic works correctly. Namely: it's a graceful shutdown, and there are no events that leak after `stopWatching` has resolved.